### PR TITLE
Lazily-initialize RegexCompiler's cached reflection members

### PR DESCRIPTION
**Pull request by stephentoub**
There are currently ~70 members that we populate to cache for later use. We rarely need all of them. This uses `field` to make them lazily initialized.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs [hunk 1]
```diff
@@ -20,88 +20,168 @@ namespace System.Text.RegularExpressions
...
-        private static readonly MethodInfo s_spanStartsWithSpanComparison = typeof(MemoryExtensions).GetMethod("StartsWith", [typeof(ReadOnlySpan<char>), typeof(ReadOnlySpan<char>), typeof(StringComparison)])!;
-        private static readonly MethodInfo s_stringAsSpanMethod = typeof(MemoryExtensions).GetMethod("AsSpan", [typeof(string)])!;
-        private static readonly MethodInfo s_stringGetCharsMethod = typeof(string).GetMethod("get_Chars", [typeof(int)])!;
-        private static readonly MethodInfo s_arrayResize = typeof(Array).GetMethod("Resize")!.MakeGenericMethod(typeof(int));
-        private static readonly MethodInfo s_mathMinIntInt = typeof(Math).GetMethod("Min", [typeof(int), typeof(int)])!;
-        private static readonly MethodInfo s_memoryMarshalGetArrayDataReferenceSearchValues = typeof(MemoryMarshal).GetMethod("GetArrayDataReference", [Type.MakeGenericMethodParameter(0).MakeArrayType()])!.MakeGenericMethod(typeof(SearchValues<char>))!;
-        private static readonly MethodInfo s_unsafeAs = typeof(Unsafe).GetMethod("As", [typeof(object)])!;
+        [field: MaybeNull]
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:23]: @MadsTorgersen, @jaredpar, @cston, we should reconsider the nullability of `field`. For these lazy init scenarios, having to put `[field: MaybeNull]` on each is a hassle and feels like it negates much of the intended simplicity of `field`.

Comment (cston) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:23]: cc @CyrusNajmabadi, @RikkiGibson

Comment (jaredpar) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:23]: The attribute approach isn't intended to be the final shipping state. There are a few options about how to approach this but none have gotten final buy off from LDM yet.

Comment (RikkiGibson) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:23]: In the upcoming cycle I am expecting we will deliver some flavor of [null-resilient getter analysis](https://github.com/dotnet/csharplang/blob/main/proposals/field-keyword.md#nullability) to figure this stuff out automatically. There were some points which we weren't able to reach consensus on last time, as well as limitations in our engineering budget.

Comment (steveharter) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:23]: For readability, perhaps include [field] on the same LOC as the `private static`.

**Review by steveharter — APPROVED** [review]

**Pull request merged**